server/src/main/java/org/opensearch/index/store/remote/file/OnDemandCompositeBlockIndexInput.java

```diff
@@ -0,0 +1,229 @@
+            OnDemandBlockIndexInput.builder()
+                .resourceDescription("OnDemandCompositeBlockIndexInput")
+                .isClone(false)
+                .offset(0L)
+                .length(remoteDirectory.fileLength(fileName)),
+            remoteDirectory,
+            fileName,
+            localDirectory,
+            fileCache,
+            context
+        );
+    }
+
+    public OnDemandCompositeBlockIndexInput(
+        Builder builder,
+        RemoteSegmentStoreDirectory remoteDirectory,
+        String fileName,
+        FSDirectory localDirectory,
+        FileCache fileCache,
+        IOContext context
+    ) throws IOException {
+        super(builder);
+        this.remoteDirectory = remoteDirectory;
+        this.localDirectory = localDirectory;
+        this.fileName = fileName;
+        this.fileCache = fileCache;
+        this.context = context;
+        originalFileSize = remoteDirectory.fileLength(fileName);
+    }
+
+    @Override
+    protected OnDemandCompositeBlockIndexInput buildSlice(String sliceDescription, long offset, long length) {
+        try {
+            return new OnDemandCompositeBlockIndexInput(
+                OnDemandBlockIndexInput.builder()
+                    .blockSizeShift(blockSizeShift)
+                    .isClone(true)
+                    .offset(this.offset + offset)
+                    .length(length)
+                    .resourceDescription(sliceDescription),
+                remoteDirectory,
+                fileName,
+                localDirectory,
+                fileCache,
+                context
+            );
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    protected IndexInput fetchBlock(int blockId) throws IOException {
+        logger.trace("fetchBlock called with blockId -> {}", blockId);
+        final String blockFileName = fileName + "_block_" + blockId;
+        final long blockStart = getBlockStart(blockId);
+        final long length = getActualBlockSize(blockId);
+        logger.trace(
+            "File: {} , Block File: {} , Length: {} , BlockSize: {} , OriginalFileSize: {}",
+            fileName,
+            blockFileName,
+            blockStart,
+            length,
+            originalFileSize
+        );
+        Path blockFilePath = getLocalFilePath(blockFileName);
+        final CachedIndexInput cacheEntry = fileCache.compute(blockFilePath, (path, cachedIndexInput) -> {
+            if (cachedIndexInput == null || cachedIndexInput.isClosed()) {
+                // Doesn't exist or is closed, either way create a new one
+                IndexInput indexInput = fetchIndexInput(blockFileName, blockStart, length);
+                return new CachedIndexInputImpl(indexInput);
+            } else {
+                logger.trace("Block already present in cache");
+                // already in the cache and ready to be used (open)
+                return cachedIndexInput;
```

Comment:
How is this going to worker when multiple threads try to read same file (or block). Dont we need to return the clone of IndexInput for each reader?

Reply:
Yes we will be cloning the IndexInputs using TransferManager. Done so in the latest revision.